File: domains/informatique/01-learning/web-development/angular/Angular/Tuto-Angular/src/app/core/services/product.service.ts
import { Injectable } from '@angular/core';
import { Observable, BehaviorSubject, tap, catchError, throwError } from 'rxjs';
import { ApiService } from './api.service';
import { Product, ProductCreate, ProductUpdate, ProductSearch } from '../models/product.model';
import { API_CONFIG } from '../constants/api.constants';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private productsSubject = new BehaviorSubject<Product[]>([]);
  public products$ = this.productsSubject.asObservable();

  private selectedProductSubject = new BehaviorSubject<Product | null>(null);
  public selectedProduct$ = this.selectedProductSubject.asObservable();

  constructor(private apiService: ApiService) {}

  // Obtenir tous les produits
  getProducts(skip: number = 0, limit: number = 100): Observable<Product[]> {
    const params = this.apiService.createParams({ skip, limit });
    
    return this.apiService.get<Product[]>(API_CONFIG.ENDPOINTS.PRODUCTS.BASE, params)
      .pipe(
        tap(products => {
          this.productsSubject.next(products);
        }),
        catchError(error => {
          console.error('Error fetching products:', error);
          return throwError(() => error);
        })
      );
  }

  // Obtenir un produit par ID
  getProduct(id: number): Observable<Product> {
    return this.apiService.get<Product>(`${API_CONFIG.ENDPOINTS.PRODUCTS.BASE}/${id}`)
      .pipe(
        tap(product => {
          this.selectedProductSubject.next(product);
        }),
        catchError(error => {
          console.error(`Error fetching product ${id}:`, error);
          return throwError(() => error);
        })
      );
  }

  // Créer un nouveau produit
  createProduct(productData: ProductCreate): Observable<Product> {
    return this.apiService.post<Product>(API_CONFIG.ENDPOINTS.PRODUCTS.BASE, productData)
      .pipe(
        tap(newProduct => {
          const currentProducts = this.productsSubject.value;
          this.productsSubject.next([...currentProducts, newProduct]);
        }),
        catchError(error => {
          console.error('Error creating product:', error);
          return throwError(() => error);
        })
      );
  }

  // Mettre à jour un produit
  updateProduct(id: number, productData: ProductUpdate): Observable<Product> {
    return this.apiService.put<Product>(`${API_CONFIG.ENDPOINTS.PRODUCTS.BASE}/${id}`, productData)
      .pipe(
        tap(updatedProduct => {
          const currentProducts = this.productsSubject.value;
          const index = currentProducts.findIndex(p => p.id === id);
          if (index !== -1) {
            currentProducts[index] = updatedProduct;
            this.productsSubject.next([...currentProducts]);
          }
          this.selectedProductSubject.next(updatedProduct);
        }),
        catchError(error => {
          console.error(`Error updating product ${id}:`, error);
          return throwError(() => error);
        })
      );
  }

  // Supprimer un produit
  deleteProduct(id: number): Observable<any> {
    return this.apiService.delete(`${API_CONFIG.ENDPOINTS.PRODUCTS.BASE}/${id}`)
      .pipe(
        tap(() => {
          const currentProducts = this.productsSubject.value;
          const filteredProducts = currentProducts.filter(p => p.id !== id);
          this.productsSubject.next(filteredProducts);
        }),
        catchError(error => {
          console.error(`Error deleting product ${id}:`, error);
          return throwError(() => error);
        })
      );
  }

  // Rechercher des produits
  searchProducts(searchParams: ProductSearch): Observable<Product[]> {
    const params = this.apiService.createParams(searchParams);
    
    return this.apiService.get<Product[]>(API_CONFIG.ENDPOINTS.PRODUCTS.SEARCH, params)
      .pipe(
        catchError(error => {
          console.error('Error searching products:', error);
          return throwError(() => error);
        })
      );
  }

  // Obtenir les produits par catégorie
  getProductsByCategory(category: string, skip: number = 0, limit: number = 100): Observable<Product[]> {
    const params = this.apiService.createParams({ skip, limit });
    
    return this.apiService.get<Product[]>(`${API_CONFIG.ENDPOINTS.PRODUCTS.BASE}/category/${category}`, params)
      .pipe(
        catchError(error => {
          console.error(`Error fetching products for category ${category}:`, error);
          return throwError(() => error);
        })
      );
  }

  // Obtenir les catégories
  getCategories(): Observable<string[]> {
    return this.apiService.get<string[]>(API_CONFIG.ENDPOINTS.PRODUCTS.CATEGORIES)
      .pipe(
        catchError(error => {
          console.error('Error fetching categories:', error);
          return throwError(() => error);
        })
      );
  }

  // Obtenir les produits en rupture de stock
  getLowStockProducts(threshold: number = 10): Observable<Product[]> {
    const params = this.apiService.createParams({ threshold });
    
    return this.apiService.get<Product[]>(API_CONFIG.ENDPOINTS.PRODUCTS.LOW_STOCK, params)
      .pipe(
        catchError(error => {
          console.error('Error fetching low stock products:', error);
          return throwError(() => error);
        })
      );
  }

  // Mettre à jour le stock
  updateStock(productId: number, quantityChange: number): Observable<Product> {
    return this.apiService.patch<Product>(
      `${API_CONFIG.ENDPOINTS.PRODUCTS.BASE}/${productId}/stock`,
      { quantity_change: quantityChange }
    ).pipe(
      tap(updatedProduct => {
        const currentProducts = this.productsSubject.value;
        const index = currentProducts.findIndex(p => p.id === productId);
        if (index !== -1) {
          currentProducts[index] = updatedProduct;
          this.productsSubject.next([...currentProducts]);
        }
      }),
      catchError(error => {
        console.error(`Error updating stock for product ${productId}:`, error);
        return throwError(() => error);
      })
    );
  }

  // Obtenir les produits actuels
  getCurrentProducts(): Product[] {
    return this.productsSubject.value;
  }

  // Obtenir le produit sélectionné
  getSelectedProduct(): Product | null {
    return this.selectedProductSubject.value;
  }

  // Effacer le cache des produits
  clearProductsCache(): void {
    this.productsSubject.next([]);
    this.selectedProductSubject.next(null);
  }
}

File: domains/informatique/01-learning/web-development/angular/Angular/Tuto-Angular/backend/services/product_service.py
from sqlalchemy.orm import Session
from typing import List, Optional
from database.models import Product
from repositories.product_repository import ProductRepository
from schemas.product_schemas import ProductCreate, ProductUpdate, Product as ProductSchema, ProductSearch
from utils.exceptions import ProductNotFoundException, InsufficientStockException

class ProductService:
    """
    Service produit - Couche Métier
    Gère la logique métier des produits
    """
    
    def __init__(self):
        self.product_repo = ProductRepository()
    
    def get_product(self, db: Session, product_id: int) -> ProductSchema:
        """
        Obtenir un produit par ID
        """
        product = self.product_repo.get(db, product_id)
        if not product:
            raise ProductNotFoundException(product_id)
        return ProductSchema.model_validate(product)
    
    def get_products(self, db: Session, skip: int = 0, limit: int = 100) -> List[ProductSchema]:
        """
        Obtenir la liste des produits avec pagination
        """
        products = self.product_repo.get_multi(db, skip=skip, limit=limit)
        return [ProductSchema.model_validate(product) for product in products]
    
    def create_product(self, db: Session, product_data: ProductCreate) -> ProductSchema:
        """
        Créer un nouveau produit
        """
        product = self.product_repo.create(db, product_data)
        return ProductSchema.model_validate(product)
    
    def update_product(self, db: Session, product_id: int, product_update: ProductUpdate) -> ProductSchema:
        """
        Mettre à jour un produit
        """
        product = self.product_repo.get(db, product_id)
        if not product:
            raise ProductNotFoundException(product_id)
        
        updated_product = self.product_repo.update(db, product, product_update)
        return ProductSchema.model_validate(updated_product)
    
    def delete_product(self, db: Session, product_id: int) -> None:
        """
        Supprimer un produit (soft delete - désactivation)
        """
        product = self.product_repo.get(db, product_id)
        if not product:
            raise ProductNotFoundException(product_id)
        
        # Soft delete - désactiver le produit au lieu de le supprimer
        product_update = ProductUpdate(is_active=False)
        self.product_repo.update(db, product, product_update)
    
    def search_products(self, db: Session, search_params: ProductSearch) -> List[ProductSchema]:
        """
        Rechercher des produits avec filtres
        """
        if search_params.query:
            products = self.product_repo.search_products(
                db, search_params.query, search_params.skip, search_params.limit
            )
        elif search_params.category:
            products = self.product_repo.get_by_category(
                db, search_params.category, search_params.skip, search_params.limit
            )
        elif search_params.min_price is not None and search_params.max_price is not None:
            products = self.product_repo.get_price_range_products(
                db, search_params.min_price, search_params.max_price, 
                search_params.skip, search_params.limit
            )
        else:
            products = self.product_repo.get_active_products(
                db, search_params.skip, search_params.limit
            )
        
        return [ProductSchema.model_validate(product) for product in products]
    
    def get_products_by_category(self, db: Session, category: str, skip: int = 0, limit: int = 100) -> List[ProductSchema]:
        """
        Obtenir les produits par catégorie
        """
        products = self.product_repo.get_by_category(db, category, skip, limit)
        return [ProductSchema.model_validate(product) for product in products]
    
    def get_categories(self, db: Session) -> List[str]:
        """
        Obtenir toutes les catégories
        """
        return self.product_repo.get_categories(db)
    
    def update_stock(self, db: Session, product_id: int, quantity_change: int) -> ProductSchema:
        """
        Mettre à jour le stock d'un produit
        """
        product = self.product_repo.get(db, product_id)
        if not product:
            raise ProductNotFoundException(product_id)
        
        # Vérifier si la réduction de stock est possible
        if quantity_change < 0 and product.stock_quantity + quantity_change < 0:
            raise InsufficientStockException(
                product.name, 
                abs(quantity_change), 
                product.stock_quantity
            )
        
        updated_product = self.product_repo.update_stock(db, product_id, quantity_change)
        return ProductSchema.model_validate(updated_product)
    
    def get_low_stock_products(self, db: Session, threshold: int = 10) -> List[ProductSchema]:
        """
        Obtenir les produits en rupture de stock
        """
        products = self.product_repo.get_low_stock_products(db, threshold)
        return [ProductSchema.model_validate(product) for product in products]
    
    def get_product_stats(self, db: Session) -> dict:
        """
        Obtenir les statistiques des produits
        """
        total_products = self.product_repo.count(db)
        active_products = len(self.product_repo.get_active_products(db, limit=10000))
        low_stock_products = len(self.product_repo.get_low_stock_products(db, 10))
        categories = self.product_repo.get_categories(db)
        
        return {
            "total_products": total_products,
            "active_products": active_products,
            "inactive_products": total_products - active_products,
            "low_stock_products": low_stock_products,
            "total_categories": len(categories),
            "categories": categories
        }
    
    def reserve_stock(self, db: Session, product_id: int, quantity: int) -> ProductSchema:
        """
        Réserver du stock pour une commande
        """
        product = self.product_repo.get(db, product_id)
        if not product:
            raise ProductNotFoundException(product_id)
        
        if product.stock_quantity < quantity:
            raise InsufficientStockException(
                product.name, quantity, product.stock_quantity
            )
        
        # Réserver le stock (réduction temporaire)
        updated_product = self.product_repo.update_stock(db, product_id, -quantity)
        return ProductSchema.model_validate(updated_product)
    
    def release_stock(self, db: Session, product_id: int, quantity: int) -> ProductSchema:
        """
        Libérer du stock réservé
        """
        product = self.product_repo.get(db, product_id)
        if not product:
            raise ProductNotFoundException(product_id)
        
        # Libérer le stock (augmentation)
        updated_product = self.product_repo.update_stock(db, product_id, quantity)
        return ProductSchema.model_validate(updated_product)
